actors/evm/src/interpreter/instructions/context.rs

```diff
@@ -59,50 +74,59 @@ pub fn coinbase<'r, BS: Blockstore, RT: Runtime<BS>>(
 
 #[inline]
 pub fn gas_price<'r, BS: Blockstore, RT: Runtime<BS>>(
-    _state: &mut ExecutionState,
-    _platform: &'r System<'r, BS, RT>,
-) -> Result<(), StatusCode> {
-    todo!("should return priority fee (needs syscall) + basefee")
+    state: &mut ExecutionState,
+    platform: &'r System<'r, BS, RT>,
+) {
+    let effective_price = platform.rt.base_fee() + platform.rt.message().gas_premium();
+    state.stack.push(U256::from(&effective_price));
+}
+
+#[inline]
+pub fn gas<'r, BS: Blockstore, RT: Runtime<BS>>(
+    state: &mut ExecutionState,
+    platform: &'r System<'r, BS, RT>,
+) {
+    state.stack.push(U256::from(platform.rt.gas_available()));
 }
 
 #[inline]
 pub fn timestamp<'r, BS: Blockstore, RT: Runtime<BS>>(
-    _state: &mut ExecutionState,
-    _platform: &'r System<'r, BS, RT>,
-) -> Result<(), StatusCode> {
-    todo!("should return the timestamp from the block header (requires syscall and FFI change)")
+    state: &mut ExecutionState,
+    platform: &'r System<'r, BS, RT>,
+) {
+    state.stack.push(U256::from(platform.rt.tipset_timestamp()));
 }
 
 #[inline]
 pub fn block_number<'r, BS: Blockstore, RT: Runtime<BS>>(
     state: &mut ExecutionState,
     platform: &'r System<'r, BS, RT>,
 ) {
-    state.stack.push(U256::from(platform.rt.curr_epoch()))
+    state.stack.push(U256::from(platform.rt.curr_epoch()));
 }
 
 #[inline]
 pub fn difficulty<'r, BS: Blockstore, RT: Runtime<BS>>(
     state: &mut ExecutionState,
     _platform: &'r System<'r, BS, RT>,
 ) {
-    state.stack.push(U256::zero())
+    state.stack.push(U256::zero());
 }
 
 #[inline]
 pub fn gas_limit<'r, BS: Blockstore, RT: Runtime<BS>>(
-    _state: &mut ExecutionState,
-    _platform: &'r System<'r, BS, RT>,
-) -> Result<(), StatusCode> {
-    todo!("requires a syscall")
+    state: &mut ExecutionState,
+    platform: &'r System<'r, BS, RT>,
+) {
+    state.stack.push(U256::from(platform.rt.message().gas_limit()));
```

Comment:
This should be the _block_ gas limit. Yeah, that's nuts. But that's what it is. We don't _really_ have an exact equivalent, but setting it to 10B is probably the right approach.

Reply:
Wait this doesnt quite make sense -- our execution is bound by the message gas limit, not the block gas limit!

Otherwise what goos is this opcode?

Reply:
I had missed this, very well spotted!

Execution is bounded by gas available, which we now return thanks to this PR in the GAS opcode. IMO this falls under the bucket of "environmental information". Maybe added to the EVM originally in expectation of variable block limits? Not sure why.

The expectation here is really to return the gas capacity of the chain's unit of tick, which in Ethereum is statically determined (single block per height). In Filecoin we have variably sized tipsets and therefore the capacity at every epoch is non-deterministic, so we're really breaking the expectation a little, but I don't think it matters in practice.

Reply:
so what do here?

Reply:
Ideally I would like us to pass these down from the client in the NetworkContext, along with the ChainID. However, for now you can hardcode and return the existing block gas limit value.

Reply:
(10B)

Reply:
done.